pkg/cache/v3/snapshot.go

```diff
@@ -81,19 +82,47 @@ func (s *Snapshot) Consistent() error {
 	if s == nil {
 		return errors.New("nil snapshot")
 	}
-	endpoints := GetResourceReferences(s.Resources[types.Cluster].Items)
-	if len(endpoints) != len(s.Resources[types.Endpoint].Items) {
-		return fmt.Errorf("mismatched endpoint reference and resource lengths: %v != %d", endpoints, len(s.Resources[types.Endpoint].Items))
-	}
-	if err := superset(endpoints, s.Resources[types.Endpoint].Items); err != nil {
-		return err
+
+	consolidated := make(map[resource.Type]map[string]bool)
+	// References made by CDS.
+	GetResourceReferencesByReference(s.Resources[types.Cluster].Items, consolidated)
```

Comment:
If we changed this function to just check all references in the snapshot I don't think we'd need to make 3 separate calls here. Does the previous consistency check not perform this same functionality? I'm a bit confused on this change

Reply:
So the old consistency check would basically do the following:

1.a) For all clusters in the snapshot, check the contents of each cluster resource, to gather if there are any references to other resources. Assume here that clusters can ONLY reference endpoints.

1.b) For each ENDPOINT in the snapshot, check to see if it is in the referenced resources list.

2.a) For all listeners in the snapshot, check the contents of each listener resource, to gather if there are any references to other resources. Assume here that listeners can ONLY reference routes.

2.b) For each ROUTE in the snapshot, check to see if it is in the referenced resources list.

Check 1.b, and 2.b here is done counterintuitively (Checking resources in the snapshot AGAINST what was referenced) vs. (Checking what was referenced AGAINST what's in the snapshot) because of a time complexity optimization. (Looking it up 1 way is O(n) vs O(1)).

Now this worked fine without SRDS. But with SRDS, listeners can reference both Scoped Routes and Routes at the same time. Which ends up breaking the entire algorithm for the following reason:

1) The "GetResourceReferences" function used to just given a list of resources, return a flat list of resources. It was then up to the caller to assume what the returned resource type was. I have updated this to instead a map looking like the following...

```
GetResourceReferences(myResources)
```
| ResourceType | Resources By Name|
| ----------- | ----------- |
| Endpoint | "endpoint1", "endpoint2" ... "endpoint N" |
| Scoped Route | "scopedroute1", "scopedroute2", ... "scopedrouteN" |
| Route | "route1", "route2", ... "route N" |

Reply:
I think with the new solution, we can do what you recommended and just check all resources for any references. I.e.

```
// Goes through the snapshot and returns a map of all referenced resources in the snapshot, keyed by type.
func GetReferencedResources() map[resource.Type]map[string]bool
```

However, we may incur some runtime tradeoffs here if we naively check every resource for references. (For instance, there's no way a route can reference another resource, so why bother checking it at all).

Reply:
Let me know what your thoughts are on this.

Reply:
I think this makes sense thanks @mtagstyle for the little write up here. So since we don't want to waste compute cycles on checking if a route is consistent, we only really need to verify listeners and their associated routes/scoped routes, clusters and their endpoints, etc.. I think I'm picking up what you're saying.

Let me go look at this new algorithm

Reply:
Okay I think this makes sense to me now

So I see why we went this way with the algorithm, I think my only hesitation at this point is it looks like the worst case scenario of time complexity here is `O(n^4)` since when we're doing the map merges we have quite a few nested loops. Are there any optimizations we can make here computation wise? I imagine at really large scale if a user checked a snapshots consistency, (10k+ configuration objects), this might cause some issues

Reply:
I'm drawing through the algorithm again, starting from "GetAllResourceReferences" and I can't see why we would have O(N^4). 

If we let N be the number of total resources in the snapshot, then I only see a static case of O(N) time complexity. To be more precise, it would be O(3*N). Refer to attached diagram for more details... Since I suck at communicating things through text.

![image](https://user-images.githubusercontent.com/22183230/135340696-b298b8c5-cb4b-4fe6-b023-4f82c6448e34.png)

As for optimizations, I think the biggest win would be to not bother checking every single resource, and only the ones we expect to have references (Refer to "resourceGroups" of Fig. 1)

Next, if we just pass the outMap all the way through to getXReferences (But add in the hacky nil checks), we would only have to pass through the referenced resources once, putting us at time complexity O(2*N).

Reply:
Let me know what you think on my recommendations for optimizations.

Reply:
Ah yes you're totally right. I went a re-read it and its not quadratic. It's definitely linear time, thanks for the diagram that helped :) 

I think it would be nice if we just go ahead and make those optimizations you recommended. In my head I imagine people making these consistency checks often (at least once per snapshot update) which would happen a lot especially on the startup of a mesh or xDS system.